boards/BMS/Src/App/App_PreChargeRelay.c

```diff
@@ -32,3 +33,32 @@ void App_PrechargeRelay_Open(const struct PrechargeRelay *const precharge_relay)
 {
     precharge_relay->open_relay();
 }
+
+bool App_PrechargeRelay_CheckFaults(
+    struct BmsCanTxInterface *can_tx,
+    bool                      is_charger_connected,
+    bool                      is_ts_rising_slowly,
+    bool                      is_ts_rising_quickly)
+{
+    static uint8_t precharge_fault_count    = 0;
+    bool           precharge_shutdown_fault = false;
+    bool           has_precharge_fault =
+        (is_charger_connected) ? is_ts_rising_slowly : (is_ts_rising_slowly | is_ts_rising_quickly);
+
+    if (has_precharge_fault)
+    {
+        precharge_fault_count++;
+    }
+    else
+    {
+        precharge_fault_count = 0;
+    }
+
+    if (precharge_fault_count >= 3)
+    {
+        precharge_shutdown_fault = true;
+    }
+
```

Comment:
Just curious, where does the 3 come from? 30ms is not much of a debounce time

Reply:
I misinterpreted the implementation, it has been reworked. See Edwin's comment on the matter for details